ADDomain fails after reboot

#### Details of the scenario you tried and the problem that is occurring
After rebooting a node the state is reported as failed. The problem is that the ADDomain resource fails with "Server instance not found".

Eventually the node will reach the state of Compliant

#### Verbose logs showing the problem

```
"Exception":  {
    "Message":  "PowerShell DSC resource MSFT_ADDomain  failed to execute Test-TargetResource functionality with error message: System.InvalidOperationException: Error getting AD domain \u0027pensam.azure\u0027. (ADD0013) ---\u003e System.ArgumentException: Server instance not found on the given port. ---\u003e System.ServiceModel.FaultException: The operation failed because of a bad parameter.\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowExceptionForFaultDetail(FaultDetail faultDetail, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowException(AdwsFault adwsFault, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.SearchAnObject(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.Search(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADWebServiceStoreAccess.Microsoft.ActiveDirectory.Management.IADSyncOperations.Search(ADSessionHandle handle, ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADObjectSearcher.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetConnectedStore()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetCmdletSessionInfo()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADGetCmdletBase`3.ADGetCmdletBaseProcessCSRoutine()\r\n   at Microsoft.ActiveDirectory.Management.CmdletSubroutinePipeline.Invoke()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.ProcessRecord()\r\n   --- End of inner exception stack trace --- ",
    "Data":  {},
    "InnerException":  {
        "ErrorRecord":  "System.InvalidOperationException: Error getting AD domain \u0027testad.azure\u0027. (ADD0013) ---\u003e System.ArgumentException: Server instance not found on the given port. ---\u003e System.ServiceModel.FaultException: The operation failed because of a bad parameter.\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowExceptionForFaultDetail(FaultDetail faultDetail, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowException(AdwsFault adwsFault, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.SearchAnObject(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.Search(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADWebServiceStoreAccess.Microsoft.ActiveDirectory.Management.IADSyncOperations.Search(ADSessionHandle handle, ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADObjectSearcher.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetConnectedStore()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetCmdletSessionInfo()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADGetCmdletBase`3.ADGetCmdletBaseProcessCSRoutine()\r\n   at Microsoft.ActiveDirectory.Management.CmdletSubroutinePipeline.Invoke()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.ProcessRecord()\r\n   --- End of inner exception stack trace ---",
        "WasThrownFromThrowStatement":  true,
        "Message":  "System.InvalidOperationException: Error getting AD domain \u0027testad.azure\u0027. (ADD0013) ---\u003e System.ArgumentException: Server instance not found on the given port. ---\u003e System.ServiceModel.FaultException: The operation failed because of a bad parameter.\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowExceptionForFaultDetail(FaultDetail faultDetail, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowException(AdwsFault adwsFault, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.SearchAnObject(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.Search(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADWebServiceStoreAccess.Microsoft.ActiveDirectory.Management.IADSyncOperations.Search(ADSessionHandle handle, ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADObjectSearcher.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetConnectedStore()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetCmdletSessionInfo()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADGetCmdletBase`3.ADGetCmdletBaseProcessCSRoutine()\r\n   at Microsoft.ActiveDirectory.Management.CmdletSubroutinePipeline.Invoke()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.ProcessRecord()\r\n   --- End of inner exception stack trace ---",
        "Data":  "System.Collections.ListDictionaryInternal",
        "InnerException":  "System.Exception: System.InvalidOperationException: Error getting AD domain \u0027testad.azure\u0027. (ADD0013) ---\u003e System.ArgumentException: Server instance not found on the given port. ---\u003e System.ServiceModel.FaultException: The operation failed because of a bad parameter.\r\n   --- End of inner exception stack trace ---\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowExceptionForFaultDetail(FaultDetail faultDetail, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.ThrowException(AdwsFault adwsFault, FaultException faultException)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.SearchAnObject(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.AdwsConnection.Search(ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADWebServiceStoreAccess.Microsoft.ActiveDirectory.Management.IADSyncOperations.Search(ADSessionHandle handle, ADSearchRequest request)\r\n   at Microsoft.ActiveDirectory.Management.ADObjectSearcher.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetRootDSE()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetConnectedStore()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.GetCmdletSessionInfo()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADGetCmdletBase`3.ADGetCmdletBaseProcessCSRoutine()\r\n   at Microsoft.ActiveDirectory.Management.CmdletSubroutinePipeline.Invoke()\r\n   at Microsoft.ActiveDirectory.Management.Commands.ADCmdletBase`1.ProcessRecord()\r\n   --- End of inner exception stack trace ---",
        "TargetSite":  "System.Collections.ObjectModel.Collection`1[System.Management.Automation.PSObject] Invoke(System.Collections.IEnumerable)",
        "StackTrace":  "   at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)\r\n   at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)\r\n   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)\r\n   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)\r\n   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)\r\n   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)\r\n   at Microsoft.PowerShell.DesiredStateConfiguration.Internal.ResourceProviderAdapter.ExecuteCommand(PowerShell powerShell, ResourceModuleInfo resInfo, String operationCmd, List`1 acceptedProperties, CimInstance nonResourcePropeties, CimInstance resourceConfiguration, LCMDebugMode debugMode, PSInvocationSettings pSInvocationSettings, UInt32\u0026 resultStatusHandle, Collection`1\u0026 result, ErrorRecord\u0026 errorRecord, PSModuleInfo localRunSpaceModuleInfo)",
        "HelpLink":  null,
        "Source":  "System.Management.Automation",
        "HResult":  -2146233087
    },
    "TargetSite":  null,
    "StackTrace":  null,
    "HelpLink":  null,
    "Source":  null,
    "HResult":  -2146233079
},
"TargetObject":  null,
"CategoryInfo":  {
    "Category":  7,
    "Activity":  "",
    "Reason":  "InvalidOperationException",
    "TargetName":  "",
    "TargetType":  ""
},
"FullyQualifiedErrorId":  "ProviderOperationExecutionFailure",
"ErrorDetails":  null,
"InvocationInfo":  null,
"ScriptStackTrace":  null,
"PipelineIterationInfo":  []
```

#### Suggested solution to the issue
Detect that the node is starting and wait for dependencies to be ready
#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
Node $AllNodes.NodeName
{

    TimeZone TimeZone { IsSingleInstance = 'Yes'; TimeZone = 'W. Europe Standard Time'}

    WaitForDisk DataDisk { DiskId = 2; RetryIntervalSec = 60; RetryCount = 60 }

    Disk DataVolume { DiskId = 2; DriveLetter = $dataDriveLetter; Size = 127.98GB; DependsOn = '[WaitForDisk]DataDisk' }

    WindowsFeature ADDS  {  Name = "AD-Domain-Services" }  

    ADDomain AD
    {
        DomainName                    = $DomainName
        Credential                    = $AdminCreds
        SafemodeAdministratorPassword = $SafeModeAdminCreds
        DomainNetBiosName             = $domainNetbiosName
        ForestMode                    = 'WinThreshold'
        DatabasePath                  = "${dataDriveLetter}:\NTDS"
        LogPath                       = "${dataDriveLetter}:\Logs"
        SysvolPath                    = "${dataDriveLetter}:\SYSVOL"
        DependsOn = '[Disk]DataVolume'
    }

    RemoteDesktopAdmin RemoteDesktopSettings
    {
        IsSingleInstance   = 'yes'
        Ensure             = 'Present'
        UserAuthentication = 'NonSecure'

        DependsOn = "[ADDomain]AD"
    }

    PendingReboot Reboot
    {
        Name = "Prior to installing tools"
        DependsOn = "[ADDomain]AD"
    }

    WaitForADDomain Ready
    {
        DomainName                    = $DomainName
        Credential                    = $AdminCreds
        DependsOn = "[PendingReboot]Reboot"
    }

    WindowsFeature RSAT 
    { 
        Ensure = "Present" 
        Name = "RSAT"

        DependsOn = "[WaitForADDomain]Ready"
    }

    WindowsFeature AdminCenter 
    { 
        Ensure = "Present" 
        Name = "RSAT-AD-AdminCenter"

        DependsOn = "[WindowsFeature]RSAT"
    }

    ADOptionalFeature RecycleBin
    {
        FeatureName                       = "Recycle Bin Feature"
        EnterpriseAdministratorCredential = $AdminCreds
        ForestFQDN                        = $DomainName
    }
}
```

#### The operating system the target node is running

```
OsName               : Microsoft Windows Server 2019 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 1809
WindowsBuildLabEx    : 17763.1.amd64fre.rs5_release.180914-1434
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
```

#### Version and build of PowerShell the target node is running

```
PSVersion                      5.1.17763.1007
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17763.1007
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

#### Version of the DSC module that was used

```
PSDesiredStateConfiguration 2.12.0
ActiveDirectoryDsc  6.0.0
ComputerManagementDsc 8.0.0
StorageDsc 4.9.0.0
```

> Hi @erictorbenson. Can you add the following verbose messages at line 113 of `MSFT_ADDomain.psm1` (within the `GetAdDomainUnexpectedError` catch block) so that we can see full details of the exception being thrown.

Done! (And I learned that `Publish-AzVMDSCConfiguration` will download the latest dependencies instead of using what you have in the folder...after running through a whole build. :-)

File is attached. Interestingly I found a positional parameter error: "[ERROR] A positional parameter cannot be found that accepts argument '+'." (I replaced the actual domain name with "redactedfqdn")

Any help would be appreciated...everything in the DSC resource looks OK so it might be something it's calling externally?

[DscExtensionHandler.0.20200406-212544.txt](https://github.com/dsccommunity/ActiveDirectoryDsc/files/4440912/DscExtensionHandler.0.20200406-212544.txt)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,12 @@ For older change log history see the [historic changelog](HISTORIC_CHANGELOG.md)
 
 ## [Unreleased]
 
+### Fixed
+
+- ADDomain
+  - Added additional Get-ADDomain retry exceptions
+    ([issue #581](https://github.com/dsccommunity/ActiveDirectoryDsc/issues/581)).
+
 ### Changed
 
 - ActiveDirectoryDsc

--- a/Tests/Unit/MSFT_ADDomain.Tests.ps1
+++ b/Tests/Unit/MSFT_ADDomain.Tests.ps1
@@ -303,6 +303,27 @@ try
                     }
                 }
 
+                Context 'When Get-ADDomain throws an ArgumentException until timeout' {
+                    BeforeAll {
+                        Mock -CommandName Get-AdDomain `
+                            -MockWith { throw New-Object -TypeName 'System.ArgumentException' }
+                        Mock -CommandName Start-Sleep
+                    }
+
+                    It 'Should throw the correct exception' {
+                        { Get-TargetResource @mockGetTargetResourceParameters } |
+                            Should -Throw ($script:localizedData.MaxDomainRetriesReachedError -f $mockDomainFQDN)
+                    }
+
+                    It 'Should call the expected mocks' {
+                        Assert-MockCalled -CommandName Get-ADDomain `
+                            -ParameterFilter { $Identity -eq $mockDomainFQDN } `
+                            -Exactly -Times $maxRetries
+                        Assert-MockCalled -CommandName Start-Sleep `
+                            -Exactly -Times $maxRetries
+                    }
+                }
+
                 Context 'When Get-ADForest throws an unexpected error' {
                     BeforeAll {
                         Mock -CommandName Get-AdForest `

--- a/source/DSCResources/MSFT_ADDomain/MSFT_ADDomain.psm1
+++ b/source/DSCResources/MSFT_ADDomain/MSFT_ADDomain.psm1
@@ -102,7 +102,8 @@ function Get-TargetResource
             }
             catch [Microsoft.ActiveDirectory.Management.ADServerDownException], `
                 [System.Security.Authentication.AuthenticationException], `
-                [System.InvalidOperationException]
+                [System.InvalidOperationException], `
+                [System.ArgumentException]
             {
                 Write-Verbose ($script:localizedData.ADServerNotReady -f $domainFQDN)
                 $domainFound = $false

